acceptance: TestGossipRestart is flaky

The following test appears to have failed:

[#9372](https://circleci.com/gh/cockroachdb/cockroach/9372):

```
I1113 22:25:34.976218 16983 acceptance/localcluster/docker.go:335  Get https://0.0.0.0:32774/_status/gossip/local: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
I1113 22:25:38.978110 16983 acceptance/localcluster/docker.go:335  Get https://0.0.0.0:32774/_status/gossip/local: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
I1113 22:25:42.980139 16983 acceptance/localcluster/docker.go:335  Get https://0.0.0.0:32774/_status/gossip/local: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
I1113 22:25:46.983590 16983 acceptance/localcluster/docker.go:335  Get https://0.0.0.0:32774/_status/gossip/local: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
I1113 22:25:47.084313 16983 acceptance/localcluster/localcluster.go:525  stopping
--- FAIL: TestGossipPeerings (48.24s)
	testing.go:157: acceptance/gossip_peerings_test.go:62: condition failed to evaluate within 20s: Get https://0.0.0.0:32774/_status/gossip/local: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
=== RUN   TestGossipRestart
I1113 22:26:00.715423 16983 acceptance/localcluster/localcluster.go:211  started docker-spy: 172.17.0.16
I1113 22:26:00.715449 16983 acceptance/localcluster/localcluster.go:218  initializing cluster
I1113 22:26:03.541642 16983 acceptance/localcluster/localcluster.go:328  creating ca (1024bit) in: /home/ubuntu/cockroach/.localcluster.certs.234859526
I1113 22:26:03.586416 16983 acceptance/localcluster/localcluster.go:333  creating node (1024bit) certs in: /home/ubuntu/cockroach/.localcluster.certs.234859526
I1113 22:26:05.845641 16983 acceptance/localcluster/localcluster.go:381  *** started roach0.local ***
  ui:    https://0.0.0.0:32776
  trace: https://0.0.0.0:32776/debug/requests
  logs:  /tmp/circle-artifacts.Hj6JOIi/acceptance/roach0.local/cockroach.INFO
--
I1113 22:27:52.012999 16983 acceptance/status_server_test.go:67  OK response from https://0.0.0.0:32797/_status/logs/3
I1113 22:27:52.017304 16983 acceptance/status_server_test.go:67  OK response from https://0.0.0.0:32797/_status/stacks/3
I1113 22:27:52.019672 16983 acceptance/status_server_test.go:67  OK response from https://0.0.0.0:32797/_status/nodes/3
I1113 22:27:52.120052 16983 acceptance/localcluster/localcluster.go:525  stopping
--- PASS: TestStatusServer (21.41s)
FAIL
ok github.com/cockroachdb/cockroach/acceptance 1337s

```
Please assign, take a look and update the issue accordingly.

--- a/gossip/client.go
+++ b/gossip/client.go
@@ -135,6 +135,7 @@ func (c *client) handleGossip(g *Gossip, call *netrpc.Call) error {
 	if call.Error != nil {
 		return call.Error
 	}
+	args := call.Args.(*Request)
 	reply := call.Reply.(*Response)
 
 	// Combine remote node's infostore delta with ours.
@@ -147,8 +148,8 @@ func (c *client) handleGossip(g *Gossip, call *netrpc.Call) error {
 				log.Infof("received %d (%d fresh) info(s) from %s", infoCount, freshCount, c.addr)
 			}
 		}
-	} else {
-		log.Infof("sent %d info(s) to %s", len(call.Args.(*Request).Delta), c.addr)
+	} else if len(args.Delta) > 0 {
+		log.Infof("sent %d info(s) to %s", len(args.Delta), c.addr)
 	}
 	c.peerID = reply.NodeID
 	g.outgoing.addNode(c.peerID)
@@ -216,16 +217,18 @@ func (c *client) gossip(g *Gossip, stopper *stop.Stopper) error {
 			if err := c.handleGossip(g, call); err != nil {
 				return err
 			}
+			req := call.Args.(*Request)
 			// If this was from a gossip pull request, fetch again.
-			if req, ok := call.Args.(*Request); ok {
-				if req.Delta == nil {
-					c.getGossip(g, addr, lAddr, done)
-				} else {
-					g.mu.Lock()
-					c.sendingGossip = false
-					g.mu.Unlock()
-					c.sendGossip(g, addr, lAddr, done)
-				}
+			if req.Delta == nil {
+				c.getGossip(g, addr, lAddr, done)
+			} else {
+				// Otherwise, it's a gossip push request; set sendingGossip
+				// flag false and maybe send more gossip if there have been
+				// additional updates.
+				g.mu.Lock()
+				c.sendingGossip = false
+				g.mu.Unlock()
+				c.sendGossip(g, addr, lAddr, done)
 			}
 		case <-c.rpcClient.Closed:
 			return util.Errorf("client closed")

--- a/gossip/client_test.go
+++ b/gossip/client_test.go
@@ -33,7 +33,7 @@ import (
 )
 
 // startGossip creates local and remote gossip instances.
-// The remote gossip instance launches its gossip service.
+// Both remote and local instances launch the gossip service.
 func startGossip(t *testing.T) (local, remote *Gossip, stopper *stop.Stopper) {
 	stopper = stop.NewStopper()
 	lclock := hlc.NewClock(hlc.UnixNano)
@@ -57,6 +57,7 @@ func startGossip(t *testing.T) (local, remote *Gossip, stopper *stop.Stopper) {
 	}); err != nil {
 		t.Fatal(err)
 	}
+
 	rclock := hlc.NewClock(hlc.UnixNano)
 	rRPCContext := rpc.NewContext(&base.Context{Insecure: true}, rclock, stopper)
 
@@ -231,7 +232,6 @@ func TestClientNodeID(t *testing.T) {
 // inbound client connection of another node.
 func TestClientDisconnectRedundant(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	t.Skip("TODO(spencer): #3350")
 	local, remote, stopper := startGossip(t)
 	defer stopper.Stop()
 	// startClient doesn't lock the underlying gossip
@@ -248,25 +248,21 @@ func TestClientDisconnectRedundant(t *testing.T) {
 	remote.mu.Unlock()
 	local.manage(stopper)
 	remote.manage(stopper)
-	wasConnected1, wasConnected2 := false, false
 	util.SucceedsWithin(t, 10*time.Second, func() error {
 		// Check which of the clients is connected to the other.
 		ok1 := local.findClient(func(c *client) bool { return c.addr.String() == rAddr.String() }) != nil
 		ok2 := remote.findClient(func(c *client) bool { return c.addr.String() == lAddr.String() }) != nil
-		if ok1 {
-			wasConnected1 = true
-		}
-		if ok2 {
-			wasConnected2 = true
-		}
-		// Check if at some point both nodes were connected to
-		// each other, but now aren't any more.
-		// Unfortunately it's difficult to get a more direct
-		// read on what's happening without really messing with
-		// the internals.
-		if wasConnected1 && wasConnected2 && (!ok1 || !ok2) {
+		// We expect node 1 to disconnect; if both are still connected,
+		// it's possible that node 1 gossiped before node 2 connected, in
+		// which case we have to gossip from node 1 to trigger the
+		// disconnect redundant client code.
+		if ok1 && ok2 {
+			if err := local.AddInfo("local-key", nil, time.Second); err != nil {
+				t.Fatal(err)
+			}
+		} else if !ok1 && ok2 {
 			return nil
 		}
-		return errors.New("not connected")
+		return errors.New("local client to remote not yet closed as redundant")
 	})
 }

--- a/gossip/gossip.go
+++ b/gossip/gossip.go
@@ -80,8 +80,8 @@ const (
 	// excessive tightening of the network.
 	minPeers = 3
 
-	// ttlNodeIDGossip is time-to-live for node ID -> address.
-	ttlNodeIDGossip = 0 * time.Second
+	// ttlNodeDescriptorGossip is time-to-live for node ID -> address.
+	ttlNodeDescriptorGossip = 0 * time.Second
 
 	// checkInterval is the default interval for checking for bad gossip
 	// states, including a badly connected network and the least useful
@@ -186,8 +186,8 @@ func (g *Gossip) SetNodeID(nodeID roachpb.NodeID) {
 // SetNodeDescriptor adds the node descriptor to the gossip network
 // and sets the infostore's node ID.
 func (g *Gossip) SetNodeDescriptor(desc *roachpb.NodeDescriptor) error {
-	log.Infof("gossiping node descriptor %+v", desc)
-	if err := g.AddInfoProto(MakeNodeIDKey(desc.NodeID), desc, ttlNodeIDGossip); err != nil {
+	log.Infof("setting node descriptor %+v", desc)
+	if err := g.AddInfoProto(MakeNodeIDKey(desc.NodeID), desc, ttlNodeDescriptorGossip); err != nil {
 		return util.Errorf("couldn't gossip descriptor for node %d: %v", desc.NodeID, err)
 	}
 	return nil
@@ -617,7 +617,7 @@ func (g *Gossip) maybeSignalStalledLocked() {
 	if g.outgoing.len()+g.incoming.len() == 0 {
 		g.signalStalled()
 	} else if g.is.getInfo(KeySentinel) == nil {
-		log.Warningf("missing sentinel gossip; assuming partition and reconnecting")
+		log.Warningf("missing sentinel gossip; assuming partition in gossip network")
 		g.signalStalled()
 	}
 }
@@ -656,6 +656,7 @@ func (g *Gossip) maybeWarnAboutInit(stopper *stop.Stopper) {
 		// This will never error because of infinite retries.
 		for r := retry.Start(retryOptions); r.Next(); {
 			g.mu.Lock()
+			hasConnections := g.outgoing.len()+g.incoming.len() > 0
 			hasSentinel := g.is.getInfo(KeySentinel) != nil
 			triedAll := g.triedAll
 			g.mu.Unlock()
@@ -663,10 +664,10 @@ func (g *Gossip) maybeWarnAboutInit(stopper *stop.Stopper) {
 			if hasSentinel {
 				break
 			}
-			// Otherwise, if all bootstrap hosts are connected, warn.
-			if triedAll {
-				log.Warningf("connected to gossip but missing sentinel. Has the cluster been initialized? " +
-					"Use \"cockroach init\" to initialize.")
+			if !hasConnections {
+				log.Warningf("not connected to gossip; check that gossip flag is set appropriately")
+			} else if triedAll {
+				log.Warningf("missing gossip sentinel; first range unavailable or cluster not initialized")
 			}
 		}
 	})
@@ -675,7 +676,8 @@ func (g *Gossip) maybeWarnAboutInit(stopper *stop.Stopper) {
 // checkHasConnected checks whether this gossip instance is connected
 // to enough of the gossip network that it has received the cluster ID
 // gossip info. Once connected, the "Connected" channel is closed to
-// signal to any waiters that the gossip instance is ready.
+// signal to any waiters that the gossip instance is ready. The gossip
+// mutex should be held by caller.
 func (g *Gossip) checkHasConnected() {
 	// Check if we have the cluster ID gossip to start.
 	// If so, then mark ourselves as trivially connected to the gossip network.
@@ -722,8 +724,8 @@ func (g *Gossip) removeClient(c *client) {
 func (g *Gossip) findClient(match func(*client) bool) *client {
 	g.clientsMu.Lock()
 	defer g.clientsMu.Unlock()
-	for i := 0; i < len(g.clients); i++ {
-		if c := g.clients[i]; match(c) {
+	for _, c := range g.clients {
+		if match(c) {
 			return c
 		}
 	}

--- a/gossip/server.go
+++ b/gossip/server.go
@@ -90,8 +90,8 @@ func (s *server) Gossip(argsI proto.Message) (proto.Message, error) {
 	// as a permanent peer. We always accept its input and return
 	// our delta.
 	canAccept := true
-	s.incoming.setMaxSize(s.maxPeers())
 	if !s.incoming.hasNode(args.NodeID) {
+		s.incoming.setMaxSize(s.maxPeers())
 		if !s.incoming.hasSpace() {
 			canAccept = false
 		} else {

